feat(cta): add rh-cta

bennypowers opened the pull request:
## What I did

1. implement rh-cta
2. add color context to `lib` (will eventually be removed from pfe, so import from here from now on)
3. add `DirController` - useful when you need to imperatively style things based on rtl

### Testing Instructions
- Examine each interaction state (default, hover, focus, active) for each CTA in https://deploy-preview-349--red-hat-design-system.netlify.app/components/cta/demo/
- `npm ci && npm run test:e2e`. Inspect the test files generated in `element/rh-cta/test/`
- @heyMP please check the integration with footer
- @zeroedin please check the integration with secondary nav. We might want to migrate secondary nav to use the local context decorators, but if this PR works, we can do that later.
- we're missing the video icon, so i didn't implement it. see discussion

## Notes to reviewers
- Please carefully review the token names used in the variants section of rh-cta.css - do all the token names used represent their semantic use? i.e. did I use border color tokens for instances of border?
- The brick-on-dark styles were not specified, so I improvised
- there is no `on saturated` as it wasn't specified
### Old Notes
This depends a lot on design-tokens, and there are lots of unanswered questions there at the time of this writing.
The biggest single feature of CTA is theming and context (which also includes `broadcast`), and it's not clear what the future holds for those concepts in both pfe and rhds. Please see the relevant discussion:
- https://github.com/RedHat-UX/red-hat-design-tokens/pull/2
- https://github.com/orgs/RedHat-UX/discussions/348
- https://github.com/orgs/RedHat-UX/discussions/306
Ensure this fixes https://github.com/RedHat-UX/red-hat-design-system/issues/238


Closes #333

bennypowers commented:
@marionnegp would you please take a preliminary look at the arrow for the default cta at https://deploy-preview-349--red-hat-design-system.netlify.app/components/cta/demo/

The change I made was to animate the `translate` value instead of the `padding` value, which is much more performant, but will perhaps change some of the pixel values, especially in relationship with sibling elements or containing elements

marionnegp commented:
I agree that it makes more sense to use `translate`! The spacing between the arrow and the text is supposed to be 8px. It looks like if you use `padding-inline: 5px;` for the arrow, it'll have the right spacing, but it's a tiny tweak.

bennypowers commented:
@RedHat-UX/designers where does the video play icon on ux.redhat.com/components/icon come from?

bennypowers commented:
If reviewers with to merge, please squash

heyMP commented:
@bennypowers thanks for all of your work on this. Looks like there is a difference in size and dark color context when rh-cta is in rh-footer.  That we'll need to investigate:

![Screen Shot 2022-08-16 at 8 22 09 PM](https://user-images.githubusercontent.com/3428964/185007582-58ef269f-8b04-4b67-9bff-3dd2b75167ad.png)

bennypowers commented:
@heyMP checking [the deploy preview](https://deploy-preview-349--red-hat-design-system.netlify.app/components/footer/demo/) as well as the dev server, the footer demo using rh-cta appears correct to me. Can you confirm? How can I reproduce your results from above?

bennypowers commented:
I removed the built-in analytics events from cta, as described in https://github.com/RedHat-UX/red-hat-design-system/issues/333#issuecomment-1209653112

The alternative is IMO much better: analytics users implement their own analytics. This way, cta users without analytics requirements do not need to load and run unnecessary javascript, and analytics users can easily capture cta click events.

I've added a demo page for analytics events, where you can peep the technique, including a javascript snippet which analytics users can copy into their analytics scripts:

https://deploy-preview-349--red-hat-design-system.netlify.app/components/cta/demo/analytics/#primary

paging @kylebuch8, @markcaron

zeroedin commented on elements/rh-cta/demo/analytics.html:
```diff
@@ -0,0 +1,76 @@
+<script type="module" src="https://unpkg.com/zero-md/dist/zero-md.min.js"></script>
+
+<p>In this demo, analytics events involving <code>&lt;rh-cta></code> elements are parsed by a 
+  document-level analytics event listener. Unlike <code>&lt;hfe-cta></code>, which implemented
```
Small typo
```Suggestion
  document-level analytics event listener. Unlike <code>&lt;pfe-cta></code>, which implemented
```

bennypowers commented on elements/rh-cta/rh-cta.css:
```diff
@@ -0,0 +1,318 @@
+  white-space: nowrap;
+  min-width: 100%;
+  color: var(--rh-cta-color);
+  font-family: var(--rh-font-family-heading, "RedHatDisplay", "Overpass", Helvetica, Arial, sans-serif);
```
see https://github.com/RedHat-UX/red-hat-design-tokens/pull/18
TL;DR: it might end up being `RedHatDisplay, "Red Hat Display"`, etc.
I'd like to write some automation that would update our components' css whenever we update the tokens, but we could do this manually as well

zeroedin commented:
I adjusted these containers and was able to get the arrow to move to its own line.

<img width="659" alt="image" src="https://user-images.githubusercontent.com/13666/185177708-402f434a-3149-4a5e-a750-3b71b9ba18d7.png">

bennypowers commented:
> I adjusted these containers and was able to get the arrow to move to its own line.
> 
> <img alt="image" width="659" src="https://user-images.githubusercontent.com/13666/185177708-402f434a-3149-4a5e-a750-3b71b9ba18d7.png">

chrome macos?

zeroedin commented:
> > I adjusted these containers and was able to get the arrow to move to its own line.
> > <img alt="image" width="659" src="https://user-images.githubusercontent.com/13666/185177708-402f434a-3149-4a5e-a750-3b71b9ba18d7.png">
> 
> chrome macos?

Yes sorry, that info would be helpful.

bennypowers commented:
alright @zeroedin, enjoy the fix for that one :wink:

zeroedin commented on elements/rh-footer/rh-global-footer.ts:
```diff
@@ -3,7 +3,7 @@ import type { ColorPalette } from '@patternfly/pfe-core';
 import { LitElement, html } from 'lit';
```
Missed import link for ColorPalette RHDS lib instead of pfe-core

```
import type { ColorPalette } from '../../lib/context/color.js';
```

bennypowers replied:
done

bennypowers commented:
:warning: This branch has been **REBASED**

bennypowers commented on elements/rh-footer/demo/rh-footer.html:
```diff
@@ -332,7 +332,7 @@ <h3 slot="header">About Red Hat</h3>
         <rh-footer-block slot="main-secondary">
           <h3 slot="header">Subscribe to our free newsletter, Red Hat Shares</h3>
-          <pfe-cta><a href="#blocks">Sign up now</a></pfe-cta>
+          <rh-cta><a href="#blocks">Sign up now</a></rh-cta>
```
Reserving this comment to block auto-merge pending @heyMP's [review](https://github.com/RedHat-UX/red-hat-design-system/pull/349#issuecomment-1217305315).

heyMP replied:
Works perfectly 👍

bennypowers commented on elements/rh-cta/demo/analytics.js:
```diff
@@ -0,0 +1,42 @@
+    const type = cta.variant ?? 'default';
+    const data = { href, text, title, color, type };
+    // eslint-disable-next-line no-console
+    console.log('CTA ANALYTICS EVENT', data);
```
Reserving this merge-blocking comment for @rmessickSDI to review the approach taken. Resolve this (and other) comments to unblock auto-merge.

bennypowers replied:
We still want analytics review and feedback but we're moving forward with the PR as is. Please open new issues for any changes needed.

zeroedin commented:
> alright @zeroedin, enjoy the fix for that one 😉

`&nbsp;` nice... 🙈

heyMP commented:
> @heyMP checking [the deploy preview](https://deploy-preview-349--red-hat-design-system.netlify.app/components/footer/demo/) as well as the dev server, the footer demo using rh-cta appears correct to me. Can you confirm? How can I reproduce your results from above?

Yeah, I don't know why I was seeing something different.  pfe-cta looks great with footer!

heyMP commented on elements/rh-cta/rh-cta.ts:
```diff
@@ -0,0 +1,148 @@
+ *
+ * @slot - We expect an anchor tag, `<a>` with an `href`, to be the first child inside `rh-cta` element. Less preferred but allowed for specific use-cases include: `<button>` (note however that the `button` tag is not supported for the default CTA styles).
+ *
+ * @fires {CtaSelectEvent} select - This event is fired when a link is clicked and serves as a way to capture click events if necessary.
+ *
```
```suggestion
```

We should remove this event as discussed here: https://github.com/RedHat-UX/red-hat-design-system/issues/333#issuecomment-1209653112

bennypowers replied:
removed

heyMP commented on lib/context/color.ts:
```diff
@@ -0,0 +1,287 @@
+}
+
+// TODO: CSS
+// 1. move sass that maps from palette to theme from _colors.scss:198+ to color-context.scss (and rename them)
+// 2. except don't because hard, wait for design tokens instead
```
😂

marionnegp reviewed (commented):
- [ ] All borders should be 1px
- [ ] When an inner border appears (for focus and active states in primary, secondary, and brick), there should be 1px of space between it in and the outer border, like this:
![brick-cta](https://user-images.githubusercontent.com/95588923/185989290-31780e3d-aa1a-4796-b4dc-c9b9b867e86d.jpg)

- [ ] Brick font-weight looks heavier than 400, but maybe that's a discrepancy between how it displays in demo and on prod.
- [ ] Video icon isn't showing up on demo.
